libbeat/monitoring/inputmon/httphandler.go

```diff
@@ -0,0 +1,162 @@
+	applicationJSON = "application/json; charset=utf-8"
+)
+
+type handler struct {
+	registry *monitoring.Registry
+}
+
+// AttachHandler attaches an HTTP handler to the given mux.Router to handle
+// requests to /inputs.
+func AttachHandler(r *mux.Router) error {
+	return attachHandler(r, globalRegistry())
+}
+
+func attachHandler(r *mux.Router, registry *monitoring.Registry) error {
+	h := &handler{registry: registry}
+	r = r.PathPrefix(route).Subrouter()
+	return r.StrictSlash(true).Handle("/", validationHandler("GET", []string{"pretty", "type"}, h.allInputs)).GetError()
+}
+
+func (h *handler) allInputs(w http.ResponseWriter, req *http.Request) {
+	requestedPretty, err := getPretty(req)
+	if err != nil {
+		serveError(w, http.StatusBadRequest, err.Error())
+		return
+	}
+
+	requestedType, err := getType(req)
+	if err != nil {
+		serveError(w, http.StatusBadRequest, err.Error())
+		return
+	}
+
+	metrics := monitoring.CollectStructSnapshot(h.registry, monitoring.Full, false)
+
+	filtered := make([]map[string]any, 0, len(metrics))
+	for _, ifc := range metrics {
+		m, ok := ifc.(map[string]any)
+		if !ok {
+			continue
+		}
+
+		if requestedType != "" {
+			if typ, ok := m["input"].(string); ok && typ != requestedType {
+				continue
+			}
+		}
+
+		filtered = append(filtered, m)
+	}
+
+	w.Header().Set(contentType, applicationJSON)
+	serveJSON(w, filtered, requestedPretty)
+}
+
+func serveJSON(w http.ResponseWriter, value any, pretty bool) {
+	w.Header().Set(contentType, applicationJSON)
+	enc := json.NewEncoder(w)
+	enc.SetEscapeHTML(false)
+	if pretty {
+		enc.SetIndent("", "  ")
+	}
+	_ = enc.Encode(value)
+}
+
+func serveError(w http.ResponseWriter, status int, txt string) {
+	w.Header().Set("Content-Type", "text/plain; charset=utf-8")
+	w.WriteHeader(status)
+	fmt.Fprintln(w, txt)
+}
+
+func getPretty(req *http.Request) (bool, error) {
+	if !req.URL.Query().Has("pretty") {
+		return false, nil
+	}
+
+	switch req.URL.Query().Get("pretty") {
+	case "", "true":
+		return true, nil
+	case "false":
+		return false, nil
+	default:
+		return false, errors.New(`invalid value for "pretty"`)
+	}
+}
+
+func getType(req *http.Request) (string, error) {
+	if !req.URL.Query().Has("type") {
+		return "", nil
+	}
+
+	switch typ := req.URL.Query().Get("type"); typ {
+	case "":
+		return "", errors.New(`"type" requires a non-empty value`)
+	default:
+		return typ, nil
+	}
+}
+
+type queryParamHandler struct {
+	allowedParams map[string]struct{}
+	next          http.Handler
+}
+
+func newQueryParamHandler(queryParams []string, h http.Handler) http.Handler {
+	m := make(map[string]struct{}, len(queryParams))
+	for _, q := range queryParams {
+		m[q] = struct{}{}
+	}
+	return &queryParamHandler{allowedParams: m, next: h}
+}
+
+func (h queryParamHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	for q := range req.URL.Query() {
+		if _, found := h.allowedParams[q]; !found {
+			serveError(w, http.StatusBadRequest, "Unknown query param "+q)
```

Comment:
As a user I prefer APIs with strict validation. If I typo a parameter in my client request then I might not catch it if the server processes my request and silently ignores my mistakes. I would rather have it fail fast up front so I can fix my errors.

And as a API maintainer I also prefer strict validation. If I want to introduce a change I don't need to consider how it might break existing misbehaving clients. For example if a client was sending some malformed request with an extra parameter that is ignored, and we want to evolve the service to interpret a parameter with that name, then it won't lead to an unexpected outcome for that client because had we prevented those requests from being accepted in the first place.

It's easier to go from strict to permissive without causing breakages for existing clients.

Reply:
Those are pretty good points.